src/lib/TopLevel.hs

```diff
@@ -528,31 +547,47 @@ execUDecl mname decl = do
     UDeclResultWorkRemaining block declAbs -> do
       result <- evalBlock block
       result' <- case declAbs of
-        Abs (ULet NoInlineLet (UPatAnn p _) _) _ ->
-          compileTopLevelFun (getNameHint p) result
+        Abs (ULet NoInlineLet (UPatAnn p _) _) _ -> do
+          ty <- getType result
+          asSpecializableFunction ty >>= \case
+            Just (n, fty) -> do
+              f <- emitBinding (getNameHint p) $
+                AtomNameBinding $ TopFunBound fty $ UnspecializedTopFun n result
+              -- warm up cache if it's already sufficiently specialized
+              when (n == 0) do
+                fSpecial <- emitSpecialization (getNameHint p) $ AppSpecialization f []
+                evalRequiredSpecializations (Var fSpecial)
+              return $ Var f
+            Nothing -> throw TypeErr $
+              "Not a valid @noinline function type: " ++ pprint ty
         _ -> return result
       emitSourceMap =<< applyUDeclAbs declAbs result'
     UDeclResultDone sourceMap' -> emitSourceMap sourceMap'
 {-# SCC execUDecl #-}
 
-compileTopLevelFun :: (Topper m, Mut n) => NameHint -> Atom n -> m n (Atom n)
-compileTopLevelFun hint f = do
-  ty <- getType f
-  naryPi <- asFirstOrderFunction ty >>= \case
-    Nothing -> throw TypeErr "@noinline functions must be first-order"
-    Just naryPi -> return naryPi
-  fSimp <- simplifyTopFunction naryPi f
+compileTopLevelFun
+  :: (Topper m, Mut n)
+  => AtomName n -> NaryPiType n -> Atom n -> [Atom n] -> m n ()
+compileTopLevelFun fname naryPi f staticArgs = do
+  f' <- forceDeferredInlining f
+  fSimp <- simplifyTopFunction naryPi f' staticArgs
+  evalRequiredSpecializations fSimp
   fImp <- toImpStandaloneFunction fSimp
-  fSimpName <- emitBinding hint $ AtomNameBinding $ SimpLamBound naryPi fSimp
-  fImpName <- emitImpFunBinding hint fImp
-  extendImpCache fSimpName fImpName
+  fImpName <- emitImpFunBinding (getNameHint fname) fImp
+  extendImpCache fname fImpName
   -- TODO: this is a hack! We need a better story for C/LLVM names.
   let cFunName = pprint fImpName
   fObj <- toCFunction cFunName fImp
   extendObjCache fImpName fObj
-  return $ Var fSimpName
 {-# SCC compileTopLevelFun #-}
 
+forceDeferredInlining :: EnvReader m => Atom n -> m n (Atom n)
```

Comment:
It's because without it, we go into an infinite loop! In `Simplify.simplifyTopFunction` we eta-expand and try to simplify `App f args`, which then recognizes that `f` is a "noinline" function and defers its simplification. But at that point we actually *do* want to do the simplification. There are other ways we could solve this, but this works well enough. I should at least add a comment.